File: hw7/tetrominos.java


public class tetrominos {
      /**
     * The tetrominos_I variable is a 2D char array that represents the 'I' shape of a tetromino.
     */
	private char[][] tetrominos_I={
        {' ',' ',' ',' '},
        {'I','I','I','I'},
        {' ',' ',' ',' '},
        {' ',' ',' ',' '},
    }; 
	private char[][] tetrominos_T={
	        {' ',' ',' ',' '},
	        {'T','T','T',' '},
	        {' ','T',' ',' '},
	        {' ',' ',' ',' '},
	};
	private char[][] tetrominos_O={
	        {' ',' ',' ',' '},
	        {'O','O',' ',' '},
	        {'O','O',' ',' '},
	        {' ',' ',' ',' '},
	};
	private char[][] tetrominos_J={
        {' ',' ',' ',' '},
        {'J',' ',' ',' '},
        {'J',' ',' ',' '},
        {'J','J',' ',' '},
    };
    private char[][] tetrominos_L={
        {' ',' ',' ',' '},
        {'L',' ',' ',' '},
        {'L',' ',' ',' '},
        {'L','L',' ',' '},
    };
    private char[][] tetrominos_Z={
        {' ',' ',' ',' '},
        {'Z','Z',' ',' '},
        {' ','Z','Z',' '},
        {' ',' ',' ',' '},
    };
    private char[][] tetrominos_S={
        {' ',' ',' ',' '},
        {' ','S','S',' '},
        {'S','S',' ',' '}, 
        {' ',' ',' ',' '},
    };
    private char typesOfTetrominos; 
    public char[][] temp;
    public int[][] position;
    public tetrominos(){};
    public tetrominos(char types){
                typesOfTetrominos=types;
    };
    public char get_typesOfTetrominos(){
        return typesOfTetrominos;
    };
    /**

    *The class tetrominos is used to create tetromino shapes and their positions.
    *It has two private variables, temp and position, which are used to store the shape and positions of the tetromino.
    *It also has several private char arrays, tetrominos_I, tetrominos_T, tetrominos_O, tetrominos_J, tetrominos_L, tetrominos_Z, tetrominos_S, which are used to store the different shapes of the tetromino.
    *It also has a private char variable, typesOfTetrominos, which is used to store the type of tetromino.

    */
    public void set_temp(){
        int i,j;
        temp=new char[4][4]; 
        for(i=0;i<4;i++){
            for(j=0;j<4;j++)
                temp[i][j] = ' ';
        }
        if(typesOfTetrominos == 'I'){
            for(i=0;i<4;i++){
                for(j=0;j<4;j++)
                    temp[i][j]=tetrominos_I[i][j];
            }
        }
        else if(typesOfTetrominos == 'T'){
            for(i=0;i<4;i++){
                for(j=0;j<4;j++)
                    temp[i][j]=tetrominos_T[i][j];
            }
        }  
        else if(typesOfTetrominos == 'O'){
            for(i=0;i<4;i++){
                for(j=0;j<4;j++)
                    temp[i][j]=tetrominos_O[i][j];
            }
        }
        else if(typesOfTetrominos == 'J'){
            for(i=0;i<4;i++){
                for(j=0;j<4;j++)
                    temp[i][j]=tetrominos_J[i][j];
            }
        }
        else if(typesOfTetrominos == 'L'){
            for(i=0;i<4;i++){
                for(j=0;j<4;j++)
                    temp[i][j]=tetrominos_L[i][j];
            }
        }
        else if(typesOfTetrominos == 'Z'){
            for(i=0;i<4;i++){
                for(j=0;j<4;j++)
                    temp[i][j]=tetrominos_Z[i][j];
            }
        }
        else if(typesOfTetrominos == 'S'){
            for(i=0;i<4;i++){
                for(j=0;j<4;j++)
                    temp[i][j]=tetrominos_S[i][j];
            }
        }
        else 
            System.out.println("Your parameter is wrong");
    };
    public char[][] getTemp(){
    	return temp;
    }

    public void set_position(){
        int i,j,k=0;
        position=new int [4][2];
        for(i=0;i<4;i++){
            for(j=0;j<2;j++)
                position[i][j] = 0;
        }
        for(i=0;i<4;i++){
            for(j=0;j<4;j++){
                if(temp[i][j]!=' '){      
                    position[k][0]=i;
                    position[k][1]=j;
                    k++;
                }
            }
        }
    };
    public int[][] getPosition(){
    	return position;
    }
    /**
 * The rotate method rotates the tetromino in a given direction and count of times.
 *
 * @param direction The direction of rotation, either "left" or "right"
 * @param count The number of times to rotate the tetromino
 */
    public void  rotate(char direction,int count){
        int i,j,a=0;
        char [][] arr=new char [4][4]; 
        for(i=0;i<4;i++){
            for(j=0;j<4;j++)
                arr[i][j] = ' ';
        }
        for( i = 0; i < 4; i++ ){
            for( j = 0; j < 4; j++ )
            	arr[i][j]=temp[i][j];
        }
        for( i = 0; i < 4; i++ ){
            for( j = 0; j < 4; j++ )
            	temp[i][j]=0;
        }
        if(direction == 'R'){
            while(a<count){
                for( i = 0; i < 4; i++ ){
                    for( j = 0; j < 4; j++ )
                    	temp[j][3-i]=arr[i][j];
                }
                a++;
            }
        }
        else{
            while(a<count){
                for( i = 0; i < 4; i++ ){
                    for( j = 0; j < 4; j++ )
                        temp[3-j][i]=arr[i][j];
                }
                a++;
            }
    }

    }
}

File: hw1/Tetrominos.cpp
#include "Tetrominos.h"


Tetrominos::Tetrominos(char types){
            typesOfTetrominos=types;
}
void Tetrominos::set_typesOfTetrominos(char types){
                typesOfTetrominos =types;
            }
char Tetrominos::get_typesOfTetrominos(){
                return typesOfTetrominos;
            }
void Tetrominos::set_temp(){
    if(typesOfTetrominos == static_cast<char>(Tetros::I))
        temp = tetrominos_I;
    else if(typesOfTetrominos == static_cast<char>(Tetros::T))
        temp = tetrominos_T;   
    else if(typesOfTetrominos == static_cast<char>(Tetros::O))
        temp = tetrominos_O;
    else if(typesOfTetrominos == static_cast<char>(Tetros::J))
        temp = tetrominos_J;
    else if(typesOfTetrominos == static_cast<char>(Tetros::L))
        temp = tetrominos_L;
    else if(typesOfTetrominos == static_cast<char>(Tetros::Z))
        temp = tetrominos_Z;
    else if(typesOfTetrominos == static_cast<char>(Tetros::S))
        temp = tetrominos_S;
    else 
        cout <<"Your parameter is wrong" << endl;
        }
vector<vector<char>> Tetrominos::get_temp(){
    return temp;
    }
void Tetrominos::print(){
    for (int i = 0; i < temp.size(); i++){
        for (int j = 0; j < temp[i].size(); j++)
            {
                cout << temp[i][j];
            }           
            cout << endl;
    }
            cout << endl;
        }
void Tetrominos::rotate(){
    int i,j;
    for( i = 0; i < 4; i++ )
    {
        for( j = 0; j < 4; j++ )
                {
                    //Clockwise rotation
                    swap(rotated_tetrominos[j][3-i],temp[i][j]);
                }
            }
            for ( i = 0; i < rotated_tetrominos.size(); i++){
                for ( j = 0; j < rotated_tetrominos[i].size(); j++)
                {
                    cout << rotated_tetrominos[i][j];
                }           
                cout << endl;
            }
            cout << endl;
        }
void Tetrominos::set_position(){
        int i,j;
    vector<int>a{0,0};
    for(i=0;i<temp.size();i++){
        for(j=0;j<temp[i].size();j++){
            if(temp[i][j]!='*'){      
                a.at(0)=i;
                a.at(1)=j;
                position.push_back(a);
            }
        }
    }
    } 
vector<vector<int>> Tetrominos::get_position(){
            return position;
        }
bool Tetrominos:: canFit(vector < vector <char>>& map){
    int i,j;
    int a=0;
        for(i=0;i<position.size();i++){
        if(map[position[i][0]][position[i][1]]=='*'){//accepting and printing non-empty indexes
            a+=1;
        }
        }
        if(a==4){//Since there are 4 full indexes, it returns the correct value when it returns 4, otherwise it gives an error.
            return true;
        }
        else{
            return false;
        } 
}   
void create_map(Tetrominos& obj,bool result,vector < vector <char>>& map){//creating map
            int i,j;
            if(result == 1){
                
                for(i=0;i<obj.get_position().size();i++)//if true the mape is printed
                    map[obj.get_position()[i][0]][obj.get_position()[i][1]]=obj.get_temp()[obj.get_position()[i][0]][obj.get_position()[i][1]];
                for(i=0;i<map.size();i++){
                    for(j=0;j<map[i].size();j++){
                        cout<<map[i][j];
                    }
                    cout<< endl;
                }
            }
            else{
                for(i=0;i<obj.get_position().size();i++){
                    obj.get_position()[i][1]+=1 ;//if false, indexes are shifted aside
                }
            }
        }